accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,194 @@
+For platform-level preview features this restriction makes sense because it
+models reality. However, for library-level preview features not only does this
+restriction not make sense, it's highly undesirable. Library authors simply want
+a way to express "these parts of my API surface might change". This statement
+isn't tied to any runtime-, framework-, or language features.
+
+Another concern we had for platform-level preview features is that we wanted to
+ensure that application developers are aware when their applications have a
+dependency on them. That is, we want to avoid a case where a library on NuGet
+uses platform-level preview features as an implementation detail that the
+application consumer isn't aware of. So we made the opt-in mechanism viral: a
+library that turns on platform-level preview features requires its consumer to
+turn them on as well.
+
+Since library-level preview features cannot change how the .NET runtime itself
+operates, we don't have that same level of concern. In fact, we believe the
+viral nature is also undesirable.
+
+## Scenarios and User Experience
+
+## Requirements
+
+### Goals
+
+* Don't change the guardrails for platform-level preview features:
+    - They require opt-in (without, the build fails with an error)
+    - They require to be viral (i.e. opt-in is also required for indirect
+      dependencies relying on preview features)
+    - The opt-in can be expressed centrally in the project file
+    - They require SDK and framework to match
+
+* Offer library-level preview features with a different set of guardrails:
+    - They require opt-in (without, the build fails with an error)
+    - The opt-in isn't viral and can be an implementation detail
+    - The opt-in is centralized, but it's specific to each preview feature
+    - Using them poses no requirements on the SDK version
+
+### Non-Goals
+
+* Relaxing or extending the guardrails for platform-level preview features
+
+## Stakeholders and Reviewers
+
+* Runtime teams
+* C#, VB, F# compiler and IDE teams
+* Library teams
+    - Core libraries
+    - ASP.NET
+    - EF
+    - Azure SDK
+    - R9
+* Project system team
+* SDK team
+* MSBuild team
+* NuGet team
+
+## Design
+
+### Attribute
+
+We'll create a new attribute, separate from the existing `System.Runtime.Versioning.RequiresPreviewFeaturesAttribute`:
+
+```C#
+namespace System.Diagnostics.CodeAnalysis;
+
+[AttributeUsage(AttributeTargets.Assembly |
+                AttributeTargets.Module |
+                AttributeTargets.Class |
+                AttributeTargets.Struct |
+                AttributeTargets.Enum |
+                AttributeTargets.Constructor |
+                AttributeTargets.Method |
+                AttributeTargets.Property |
+                AttributeTargets.Field |
+                AttributeTargets.Event |
+                AttributeTargets.Interface |
+                AttributeTargets.Delegate, Inherited = false)]
+public sealed class PreviewAttribute : Attribute
```

Comment:
I'm open to using a the term *experimental* in order to differentiate from the existing `RequiresPreviewFeaturesAttribute`. The downside is that we generally don't use that terminology because and instead use *preview*.

Having two different attributes in two different namespaces should be enough to differentiate their use.

* `System.Runtime.Versioning` is basically not used by anyone; I don't think people are like to discover it.
* The name `RequiresPreviewFeatures` is a bit of a mouthful, so I don't think it's likely that people type it out and accidentally complete the using suggestion from the IDE either.
* The namespace `System.Diagnostics.CodeAnalysis` has become the standard namespace for API-specific annotations, like for nullable reference types.
* The name `PreviewAttribute` is quite short and attractive, so is more likely to be discovered by library authors first.

Reply:
Hmm.

```
[Preview]
void Foo();

[Experimental]
void Foo();
```

When I see [Preview], it doesn't strike me as being descriptive and feels more imperative. "Show me a preview" kind of thing. [Experimental] doesn't have that problem. I think the equivalent form using preview-based terminology would be things like:

```
[InPreview]
void Foo();

[PreviewFeature]
void Foo();

[PreviewApi]
void Foo();
```

But I still prefer [Experimental] :-)

Reply:
Is it because `Preview` isn't an adjective?

Reply:
Is it possible some may read preview as its common sense of "first look at something unfinished that will be stable later"? if so, Experimental may have the advantage that it doesn't imply the feature is going to stay.

Reply:
> experimental in order to differentiate from the existing RequiresPreviewFeaturesAttribute

I think differentiation would be valuable here given the conceptual overlap

> . The downside is that we generally don't use that terminology and instead use preview.

Microsoft use that term, but library authors will be applying this attribute. In my experience it is typical for OSS libraries to refer to features that may change as “experimental.”

Reply:
WinRT uses the name [`ExperimentalAttribute`](https://learn.microsoft.com/en-us/uwp/api/Windows.Foundation.Metadata.ExperimentalAttribute?view=winrt-22621) for this scenario.

Reply:
Looks we're leaning towards `Experimental` then.

Reply:
Resolved with e8797e507eb45df346ca66c112f570bf61a90782.